### ISSUE TITLE
Building/Testing for Android

### ISSUE THREAD

**Issue by tommywalkie** [tommywalkie/sample-godot-rust-app#12]
**Current roadmap**
- [x] Compile Rust source (`gdnative`, `core`, etc.) for Android targets
  - [x] Build for **ARMv7** using `armv7-linux-androideabi` toolchain
  - [x] Build for **AArch64** using `aarch64-linux-android` toolchain
  - [x] Build for **i686** using `i686-linux-android` toolchain
- [x] ~~Make the tests work when using `cargo test`~~ See the following [statement](https://github.com/tommywalkie/sample-godot-rust-app/issues/12#issuecomment-616384481) for details.
- [x] Export Cargo release builds for **ARMv7**, **AArch64** and **i686** as artifacts 
- [x] Build an app with GUI and install the game on a device and see if it works
- [x] Setup headless Godot Engine instance to export games for Android
  - [x] See if it is possible to set **jarsigner**, `adb` and **debug keystore** paths (yes, [it is](https://github.com/tommywalkie/sample-godot-rust-app/issues/12#issuecomment-616508132)) 
  - [x] Build an app with Godot CLI
  - [x] Export, install the game on a device and see if it works
- [x] Cleaning / Optimizing the Github Actions workflow
- [x] Write docs about how the whole thing could be done

**Comment by tommywalkie** [comment]
So far, mostly thanks to [@grzi](https://github.com/grzi), I could finally manage to build the app with `armv7-linux-androideabi` toolchain.
Except the integration tests are failing when launching Cargo tests `C_INCLUDE_PATH=.:$JAVA_HOME/include/:$JAVA_HOME/include/linux/ cargo test --release --target armv7-linux-androideabi` :
```
   Compiling speculate v0.1.2
   Compiling core v0.1.0 (/home/runner/work/sample-godot-rust-app/sample-godot-rust-app/src/core)
    Finished release [optimized] target(s) in 44.48s
     Running target/armv7-linux-androideabi/release/deps/core-465f67e7f26363c4
/home/runner/work/sample-godot-rust-app/sample-godot-rust-app/target/armv7-linux-androideabi/release/deps/core-465f67e7f26363c4: 1: /home/runner/work/sample-godot-rust-app/sample-godot-rust-app/target/armv7-linux-androideabi/release/deps/core-465f67e7f26363c4: Syntax error: word unexpected (expecting ")")
error: test failed, to rerun pass '-p core --lib'
##[error]Process completed with exit code 2.
```
However tests are fine for Windows and Linux.
One possible reason why this is happening is because ARMv7 compiled libraries cannot be correctly read on Ubuntu. Need to do some research about that.

That said, after more than 50 commits, a successful build is good news.

**Comment by tommywalkie** [comment]
Small update, _looks like_ I was correct about ARMv7 vs x86/x64 ([source](https://www.walkerb.net/blog/x64-vs-arm/))

> The two processor architectures are mutually exclusive: a program that’s built for x86/x64 can’t run on ARM under any circumstances, and vice versa.

I'm assuming i can skip Rust tests step for Android target and state that building for ARMv7 did something that was expected, not working on Ubuntu 😅

That said, might be good to **move Rust tests to its own single Github Actions job** and be the first job to run. If it fails, there is no point trying to build for x86/x64, Windows, ARM, MacOS or whatever target.

For the Android target part, I may need to test the ARM based `libcore.so` in a local project, see if the APK works and then decide of what comes next. I'll _probably_ run into a blocking issue when setting up the CI part for releasing Android app since [Godot CLI docs doesn't mention a way to edit engine settings](https://docs.godotengine.org/en/3.2/getting_started/editor/command_line_tutorial.html).

The reason why I would need to edit some engine settings is to set **jarsigner**, `adb` and **debug keystore** so basically everything that Godot Engine GUI teels us what it needs in order to build for Android.

![image](https://user-images.githubusercontent.com/16148332/79728319-0a9b5200-82ee-11ea-9c3a-dd4be1eebe35.png)

**Comment by tommywalkie** [comment]
For reference, this is what I did to make the build work / what needs to be optimized on [`build-for-android-from-ci`](https://github.com/tommywalkie/sample-godot-rust-app/tree/build-for-android-from-ci) branch :

- Use `armv7-linux-androideabi` toolchain
```bash
rustup target add armv7-linux-androideabi
```
- Install the `llvm-config` executable that is needed by `gdnative-sys`. By default, [Github's Ubuntu machine](https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1804-README.md) has a LLVM prebuilt instance but they are not providing a `llvm-config` executable but _something_ here in this overkill command is providing it.
```bash
sudo apt-get install clang-format clang-tidy clang-tools clang libc++-dev libc++1 libc++abi-dev libc++abi1 libclang-dev libclang1 libomp-dev libomp5 lld lldb llvm-dev llvm-runtime llvm -y
```
- Enable cross-compilation with standard C and C++ libraries ([source](https://askubuntu.com/a/855957)).
```bash
sudo apt-get update
sudo apt-get install g++-multilib gcc-multilib libc6-dev-i386 -y
```
- Edit `~/.cargo/config` file and use Ubuntu based Github machine's NDK as a `cc` linker.
```bash
ls ~/.cargo/  
touch ~/.cargo/config
echo '[target.armv7-linux-androideabi]' >> ~/.cargo/config
echo 'ar = "/usr/local/lib/android/sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-ar"' >> ~/.cargo/config
echo 'linker = "/usr/local/lib/android/sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi29-clang"' >> ~/.cargo/config
cat ~/.cargo/config
```
- Finally, build with Cargo and link it to the JNI from the Ubuntu based Github machine's JDK. `jni.h` is located in `$JAVA_HOME/include/` and requires `jni_md.h` which is located in `$JAVA_HOME/include/linux/` 
```bash
C_INCLUDE_PATH=.:$JAVA_HOME/include/:$JAVA_HOME/include/linux/ cargo build --release --target armv7-linux-androideabi
```

And the build should work.
```bash
...
Compiling gdnative-sys v0.8.0
Compiling gdnative-core v0.8.0
Compiling gdnative-bindings v0.8.0
Compiling gdnative v0.8.0
Compiling fullscreen_colored_panel v0.1.0 (/home/runner/work/sample-godot-rust-app/sample-godot-rust-app/src/fullscreen_colored_panel)
Compiling core v0.1.0 (/home/runner/work/sample-godot-rust-app/sample-godot-rust-app/src/core)
Finished release [optimized] target(s) in 3m 52s
```

**Comment by tommywalkie** [comment]
Update for `aarch64` :
Using proper `cc` linker for `aarch64`, looks like the same steps worked with `aarch64-linux-android` toolchain. All the previously mentionned [steps](https://github.com/tommywalkie/sample-godot-rust-app/issues/12#issuecomment-616408027) might not be necessary (especially the cross-compilation part) but it works on [`build-for-android-from-ci`](https://github.com/tommywalkie/sample-godot-rust-app/tree/build-for-android-from-ci) branch. ✅
```bash
C_INCLUDE_PATH=.:$JAVA_HOME/include/:$JAVA_HOME/include/linux/ cargo build --release --target aarch64-linux-android
   ...
   Compiling gdnative-sys v0.8.0
   Compiling gdnative-core v0.8.0
   Compiling gdnative-bindings v0.8.0
   Compiling gdnative v0.8.0
   Compiling fullscreen_colored_panel v0.1.0 (/home/runner/work/sample-godot-rust-app/sample-godot-rust-app/src/fullscreen_colored_panel)
   Compiling core v0.1.0 (/home/runner/work/sample-godot-rust-app/sample-godot-rust-app/src/core)
    Finished release [optimized] target(s) in 4m 04s
```

**Comment by tommywalkie** [comment]
Just noticed Godot Engine editor settings are saved in a `editor_settings-3.tres` config file.

> Three virtual methods are added to OS for the various XDG paths we will use:
>  OS::get_data_path gives XDG_DATA_HOME, or if missing:
    ~/.local/share on X11, ~/Library/Application Support/ on macOS and %APPDATA% on Windows
    OS::get_config_path gives XDG_CONFIG_HOME, or if missing:
    ~/.config on X11, ~/Library/Application Support/ on macOS and %APPDATA% on Windows
    OS::get_cache_path gives XDG_CACHE_HOME, or if missing:
    ~/.cache on X11, ~/Library/Caches on macOS and %TEMP% on Windows

The good news is we actually can edit **jarsigner**, `adb` and **debug keystore** paths.
```
export/android/adb = ""
export/android/jarsigner = ""
export/android/debug_keystore = ""
export/android/debug_keystore_user = ""
export/android/debug_keystore_pass = ""
```

**TL;DR** : Building Godot game for Android via Github Actions should be possible.

**Comment by tommywalkie** [comment]
**First test on Android** — Using a Xiaomi Mi A1 device.

1. Compiled the app via Godot Engine GUI with the ARMv7 based `libcore.so` as an entry, and with correct jarsigner, `adb` and debug keystore paths, using the `editor_settings-3.tres` file.

2. Installed the app on my device, the first thing we can notice is the app actually launches without crashing and displays Godot Engine splash screen.

![Screenshot_20200420-142809](https://user-images.githubusercontent.com/16148332/79752053-f407f180-8313-11ea-8a9c-cb46f8daf103.png)

3. The actual `sample_godot_rust_app` appears. First scene **background screen should be green but instead is grey** (in the context of the app, background color is handled via Rust/GDNative). The GDScript based button "_Go to second scene_" works and redirects on the second scene. **However the Rust/GDNative based button** "_Go to first scene_" **doesn't work** but the app doesn't crash.

![Screenshot_20200420-141854](https://user-images.githubusercontent.com/16148332/79752506-b35ca800-8314-11ea-9a72-88a3ab56ed6c.png)

The main thing we can learn from this test is GDNative based scripts need something else in order to work. Maybe need to dig deeper into ["Using Godot's GDNative on Android"](https://www.sisik.eu/blog/godot/gd-native-on-android) blog post.

**Comment by tommywalkie** [comment]
**Second test on Android** — Using a Xiaomi Mi A1 device and a x86 Nexus based AVD.

1. I might have misunderstood which architecture was using my Xiaomi device so I added support for more architectures on `scenes/core.gdnlib`.

```diff
[entry]

X11.64="res://target/stable-x86_64-unknown-linux-gnu/release/libcore.so"
OSX.64="res://target/release/core.dylib"
Windows.64="res://target/x86_64-pc-windows-gnu/release/core.dll"
Android.armeabi-v7a="res://target/armv7-linux-androideabi/release/libcore.so"
+ Android.arm64-v8a="res://target/aarch64-linux-android/release/libcore.so"
+ Android.x86="res://target/i686-linux-android/release/libcore.so"
+ Android.x86_64="res://target/aarch64-linux-android/release/libcore.so"

[dependencies]

X11.64=[  ]
OSX.64=[  ]
Android.armeabi-v7a=[  ]
+ Android.arm64-v8a=[  ]
+ Android.x86=[  ]
+ Android.x86_64=[  ]

[general]

singleton=false
load_once=true
symbol_prefix="godot_"
reloadable=true
```

2. Set `true` the `x86` and `x86_64` support, on `export_presets.cfg` , which was `false` by default.

```diff
architectures/armeabi-v7a=true
architectures/arm64-v8a=true
- architectures/x86=false
- architectures/x86_64=false
+ architectures/x86=true
+ architectures/x86_64=true
```

3. Re-exported the Android game. Tested both with x86 Nexus based AVD and my Xiaomi M1 A1 device and well, it works ! 🎉

![20200420_165048](https://user-images.githubusercontent.com/16148332/79767191-46074200-8329-11ea-807d-91e6ad046286.gif)

**Comment by tommywalkie** [comment]
**Update on Android export via Github Actions** :
I believe the tools are there, we can edit paths for **jarsigner**, `adb` and `debug.keystore` (which can be generated via `keytool` command), the Godot CLI tool works fine for Windows and Linux exports, but I still encounter errors while trying to export for Android, using a headless Godot Engine instance on Ubuntu, while doing it from my personal Windows machine is 100% working.
```
godot project.godot -q --export "Android" ./build/android/sample_godot_rust_app.apk
        ...
	export: step 103: Signing release APK...
ERROR: _fs_changed: Project export failed with error code 12 for preset 'Android'.
export: end
   At: editor/editor_node.cpp:634.
reimport: end
ERROR: ~List: Condition "_first != __null" is true.
   At: ./core/self_list.h:112.
ERROR: cleanup: There are still MemoryPool allocs in use at exit!
   At: core/pool_vector.cpp:69.
```
I'm suspecting the **jarsigner** step is failing but not sure how to prove it.
Looks like it is related to https://github.com/godotengine/godot/issues/29385, I can try building [`bemyak/godot-ci`](https://github.com/bemyak/godot-ci) Docker image. Until a suitable setup is found, you can still download artifacts from Cargo builds.

![image](https://user-images.githubusercontent.com/16148332/79870966-05203380-83e4-11ea-8ad3-586c1c1549e2.png)

And then build a Android app using your own Godot Engine setup, as long as the `target` folder structure looks like this (otherwise, edit `scenes/core.gdnlib` at your convenience) :

```
.target
├── aarch64-linux-android
│   └── release
│       └── libcore.so
├── armv7-linux-androideabi
│   └── release
│       └── libcore.so
├── i686-linux-android
│   └── release
│       └── libcore.so
├── stable-x86_64-unknown-linux-gnu
│   └── release
│       └── libcore.so
└── x86_64-pc-windows-gnu
     └── release
          └── core.dll
```

**Comment by Calinou** [comment]
> I'll probably run into a blocking issue when setting up the CI part for releasing Android app since Godot CLI docs doesn't mention a way to edit engine settings.

[With `sed`, anything is possible](https://github.com/Calinou/escape-space/blob/e5f2f63fec6ef291e7c18d5b4d21f78e811a4fe2/.azure-pipelines/build.sh#L53-L58) :stuck_out_tongue: 

But yes, there's no easy way to override project or editor settings from the command-line yet. I tried to implement it a few months ago but couldn't get it to work.

**Comment by tommywalkie** [comment]
@Calinou The funny thing is I did try with [`echo >>` commands](https://github.com/tommywalkie/sample-godot-rust-app/blob/master/.github/workflows/godot-android.yml#L167-L173) in master branch and then, in current working branch, a Dockerfile borrowed from `godot-ci` which uses [`sed` commands](https://github.com/tommywalkie/sample-godot-rust-app/blob/better-build-for-android/dockerfiles/godot-ci.Dockerfile#L21-L31), I guess these methods both work but they still lead to the [last issue mentionned](https://github.com/tommywalkie/sample-godot-rust-app/issues/12#issuecomment-617186002).

**Comment by tommywalkie** [comment]
**Final update** :
As of https://github.com/tommywalkie/sample-godot-rust-app/pull/15, the Ci workflow is now ready to export GDNative/Rust based Godot games for Android targets. Using an encrypted keystore and two Github secrets for the keystore password and the GPG key passphrase. This should be the current most trivial way to export games for Android via Github Actions.
Closing the issue.
